File: android-app/app/src/main/java/com/servicos/app/ui/screens/TelaInicial.kt
@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package com.servicos.app.ui.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.servicos.app.data.DadosMock
import com.servicos.app.model.Categoria
import com.servicos.app.ui.components.PrestadorCard
import com.servicos.app.ui.theme.Verde

/** Tela de abertura: busca, grade de categorias e prestadores bem avaliados. */
@Composable
fun TelaInicial(
    onCategoria: (String) -> Unit,
    onPrestador: (String) -> Unit,
) {
    val destaques = DadosMock.destaques()

    Column {
        TopAppBar(
            title = { Text("O que você precisa hoje?") },
            colors = TopAppBarDefaults.topAppBarColors(
                containerColor = Verde,
                titleContentColor = Color.White,
            ),
        )

        LazyColumn(contentPadding = PaddingValues(bottom = 24.dp)) {
            // Barra de busca (ainda visual)
            item {
                OutlinedTextField(
                    value = "",
                    onValueChange = {},
                    readOnly = true,
                    placeholder = { Text("Buscar serviço ou profissional...") },
                    leadingIcon = { Icon(Icons.Filled.Search, contentDescription = null) },
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp),
                )
            }

            item { TituloSecao("Categorias") }

            // Grade de categorias em linhas de 4
            item {
                Column(modifier = Modifier.padding(horizontal = 12.dp)) {
                    DadosMock.categorias.chunked(4).forEach { linha ->
                        Row(modifier = Modifier.fillMaxWidth()) {
                            linha.forEach { categoria ->
                                Box(modifier = Modifier.weight(1f)) {
                                    CategoriaBotao(
                                        categoria = categoria,
                                        onClick = { onCategoria(categoria.id) },
                                    )
                                }
                            }
                            // completa a última linha para manter o alinhamento
                            repeat(4 - linha.size) {
                                Spacer(modifier = Modifier.weight(1f))
                            }
                        }
                    }
                }
            }

            item { TituloSecao("Bem avaliados") }

            items(destaques, key = { it.id }) { prestador ->
                PrestadorCard(
                    prestador = prestador,
                    onClick = { onPrestador(prestador.id) },
                )
            }
        }
    }
}

@Composable
private fun TituloSecao(texto: String) {
    Text(
        text = texto,
        style = MaterialTheme.typography.titleLarge,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.padding(start = 16.dp, end = 16.dp, top = 12.dp, bottom = 4.dp),
    )
}

/** Um ícone de categoria clicável na grade. */
@Composable
private fun CategoriaBotao(categoria: Categoria, onClick: () -> Unit) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier
            .padding(4.dp)
            .clip(RoundedCornerShape(14.dp))
            .clickable(onClick = onClick)
            .padding(4.dp),
    ) {
        Card(
            shape = RoundedCornerShape(16.dp),
            colors = CardDefaults.cardColors(containerColor = Color.White),
            modifier = Modifier.size(56.dp),
        ) {
            Box(contentAlignment = Alignment.Center, modifier = Modifier.fillMaxWidth().height(56.dp)) {
                Icon(categoria.icone, contentDescription = null, tint = Verde, modifier = Modifier.size(26.dp))
            }
        }
        Spacer(Modifier.height(4.dp))
        Text(
            text = categoria.nome,
            style = MaterialTheme.typography.labelSmall,
            textAlign = TextAlign.Center,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

File: app/lib/screens/tela_inicial.dart
import 'package:flutter/material.dart';

import '../data/dados_mock.dart';
import '../models/categoria.dart';
import '../models/prestador.dart';
import '../theme/tema.dart';
import '../widgets/prestador_card.dart';
import 'tela_categoria.dart';
import 'tela_prestador.dart';

/// Tela de abertura: saudação, busca, grade de categorias e destaques.
class TelaInicial extends StatelessWidget {
  const TelaInicial({super.key});

  @override
  Widget build(BuildContext context) {
    // Destaques = prestadores mais bem avaliados no geral.
    final destaques = [...DadosMock.prestadores]
      ..sort((a, b) => b.mediaEstrelas.compareTo(a.mediaEstrelas));

    return Scaffold(
      body: CustomScrollView(
        slivers: [
          // Topo verde com título e barra de busca
          SliverAppBar(
            pinned: true,
            expandedHeight: 130,
            backgroundColor: Tema.verde,
            flexibleSpace: FlexibleSpaceBar(
              titlePadding:
                  const EdgeInsets.only(left: 16, bottom: 16, right: 16),
              title: const Text(
                'O que você precisa hoje?',
                style: TextStyle(fontSize: 16),
              ),
            ),
          ),

          SliverToBoxAdapter(
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: _BarraBusca(),
            ),
          ),

          // Título da seção de categorias
          const SliverToBoxAdapter(
            child: Padding(
              padding: EdgeInsets.fromLTRB(16, 0, 16, 8),
              child: Text('Categorias',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
            ),
          ),

          // Grade de categorias
          SliverPadding(
            padding: const EdgeInsets.symmetric(horizontal: 12),
            sliver: SliverGrid(
              gridDelegate:
                  const SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 4,
                mainAxisSpacing: 4,
                crossAxisSpacing: 4,
                childAspectRatio: 0.85,
              ),
              delegate: SliverChildBuilderDelegate(
                (context, i) => _CategoriaBotao(
                  categoria: DadosMock.categorias[i],
                ),
                childCount: DadosMock.categorias.length,
              ),
            ),
          ),

          // Título da seção de destaques
          const SliverToBoxAdapter(
            child: Padding(
              padding: EdgeInsets.fromLTRB(16, 16, 16, 4),
              child: Text('Bem avaliados',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
            ),
          ),

          // Lista de destaques
          SliverList(
            delegate: SliverChildBuilderDelegate(
              (context, i) {
                final Prestador p = destaques[i];
                return PrestadorCard(
                  prestador: p,
                  onTap: () => Navigator.of(context).push(
                    MaterialPageRoute(
                      builder: (_) => TelaPrestador(prestador: p),
                    ),
                  ),
                );
              },
              childCount: destaques.length,
            ),
          ),

          const SliverToBoxAdapter(child: SizedBox(height: 24)),
        ],
      ),
    );
  }
}

/// Barra de busca (ainda visual — a busca real entra numa próxima etapa).
class _BarraBusca extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextField(
      decoration: InputDecoration(
        hintText: 'Buscar serviço ou profissional...',
        prefixIcon: const Icon(Icons.search),
        filled: true,
        fillColor: Colors.white,
        contentPadding: const EdgeInsets.symmetric(vertical: 0),
        border: OutlineInputBorder(
          borderRadius: BorderRadius.circular(14),
          borderSide: BorderSide.none,
        ),
      ),
      onSubmitted: (_) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Busca entra na próxima etapa 🙂')),
        );
      },
    );
  }
}

/// Um ícone de categoria clicável na grade.
class _CategoriaBotao extends StatelessWidget {
  final Categoria categoria;

  const _CategoriaBotao({required this.categoria});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      borderRadius: BorderRadius.circular(14),
      onTap: () => Navigator.of(context).push(
        MaterialPageRoute(
          builder: (_) => TelaCategoria(categoria: categoria),
        ),
      ),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Container(
            width: 54,
            height: 54,
            decoration: BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.circular(16),
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withOpacity(0.05),
                  blurRadius: 6,
                  offset: const Offset(0, 2),
                ),
              ],
            ),
            child: Icon(categoria.icone, color: Tema.verde, size: 26),
          ),
          const SizedBox(height: 4),
          Text(
            categoria.nome,
            textAlign: TextAlign.center,
            style: const TextStyle(fontSize: 11),
            maxLines: 2,
            overflow: TextOverflow.ellipsis,
          ),
        ],
      ),
    );
  }
}
